app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -45,46 +45,41 @@ private val EVENTS = listOf(
     private val app: Application,
 ) {
-    private var torService: OnionService? = null
     private var broadcastReceiver: BroadcastReceiver? = null
-    private val serviceConnection = object : ServiceConnection {
-        override fun onServiceConnected(className: ComponentName, service: IBinder) {
-            LOG.info("OnionService connected")
-            val binder = service as OnionService.OnionBinder
-            torService = binder.service
-        }
-
-        override fun onServiceDisconnected(arg0: ComponentName) {
-            LOG.info("OnionService disconnected")
-            torService = null
-        }
-    }
 
     /**
      * Starts [TorService] and creates a new onion service.
      * Suspends until the address of the onion service is available.
      */
-    suspend fun start(port: Int): String = suspendCoroutine { continuation ->
+    suspend fun start(port: Int): String = suspendCancellableCoroutine { continuation ->
+        LOG.info("Starting...")
         broadcastReceiver = object : BroadcastReceiver() {
             override fun onReceive(context: Context, i: Intent) {
                 when (i.getStringExtra(EXTRA_STATUS)) {
                     TorService.STATUS_STARTING -> LOG.debug("TorService: Starting...")
                     TorService.STATUS_ON -> {
-                        // we are receiving a broadcast from the service, it must be set
-                        val service = torService!!
                         LOG.debug("TorService: Started")
-                        try {
-                            // the control connection is only guaranteed to be available here
-                            // see: https://github.com/guardianproject/tor-android/issues/56
-                            service.torControlConnection.setEvents(EVENTS)
-                            service.torControlConnection.addRawEventListener { keyword, data ->
+
+                        startControlConnection(context).apply {
+                            try {
+                                launchThread(true)
+                                authenticate(ByteArray(0))
+                                takeOwnership()
+                                setEvents(EVENTS)
+                            } catch (e: Exception) {
+                                // gets caught and logged by caller
+                                continuation.resumeWithException(e)
+                                return
+                            }
+                            val onion = createOnionService(this, continuation, port)
+                            // if onion is null, we resume with an exception in createOnionService()
+                            if (onion != null) addRawEventListener { keyword, data ->
                                 LOG.debug("$keyword: $data")
+                                if (keyword == "HS_DESC" && data.startsWith("UPLOADED ")) {
+                                    if (continuation.isActive) continuation.resume(onion)
```

Comment:
Should we remove the listener after the first upload event?

Reply:
Depends. If we don't want/need to see the Tor logs, we could.

Reply:
Is it a problem if the listener keeps a reference to the continuation and keeps calling resume() on it after the continuation would normally have finished? Hopefully that's not a memory leak?

Reply:
Oh sorry, just saw in the other thread that you're removing the listener now. Never mind.